### DISCUSSION TITLE
Typing Promise returning function with Result

### DISCUSSION THREAD

**Discussion by kee-oth** [true-myth/true-myth#430] [Q&A] [answered]
Hi!

I'm looking for help with how best to type a function like this:

```TS
const getArtist
  = (identifier) => fetchArtist (`/artists/${identifier}`)
      .then ((response: GetArtistResponse) => Result.ok (response.data))
      .catch (() => Result.err (ErrorMessage.ArtistNotFound))
```

I'd like to do something compact like this but TypeScript doesn't like it:
```TS
const getArtist
  : (_:string) =>  Promise<Result<Artist, ErrorMessage>>
  = (identifier) => fetchArtist (`/artists/${identifier}`)
    .then ((response: GetArtistResponse) => Result.ok (response.data))
    .catch (() => Result.err (ErrorMessage.ArtistNotFound))
```

TypeScript is happy with this though:
```TS
const getArtist
  : (_:string) =>  Promise<Ok<Artist> | Err<ErrorMessage>>
  = (identifier) => fetchArtist (`/artists/${identifier}`)
    .then ((response: GetArtistResponse) => Result.ok (response.data))
    .catch (() => Result.err (ErrorMessage.ArtistNotFound))
```

Is this just the way it has to be?

Thanks!

**Comment by chriskrycho** [comment, marked as answer]
Hi @kee-oth, thanks for the ask. You will need to provide the type for the initial parameter *somehow*—this isn’t a True Myth limitation but a TypeScript limitation. Unlike languages like F♯, OCaml, Elm, Haskell, etc., TypeScript does not do Hindley-Milner inference or any other kind of full program inference; any function parameter which is not explicitly typed in *some* way is implicitly typed as `any`. (I say “in *some* way” because callbacks can get their types from the context where they’re used, but the definition of the site they are *passed* to has to be typed well for that to work.) The best thing you can do is something like your final version *or* simply providing the type explicitly inline.

A couple other notes:

1. It's idiomatic to type it as `Result<T, E>` rather than `Ok<T> | Err<E>`.
2. Assuming `fetchArtist` returns `GetArtistResponse` when successful, you *don’*t need the type annotation on that callback.

Net, I would write this like this:

```ts
const getArtist = (identifier: string): Promise<Result<Artist, ErrorMessage>> =>
  fetchArtist(`/artists/${identifier}`)
    .then((response) => Result.ok(response.data))
    .catch(() => Result.err(ErrorMessage.ArtistNotFound))
```

**Reply by franklin-huang** [reply]
I'm facing the same issue and the suggested solution is not quite working for me tho. Here is the screen shot of my code:
![image](https://user-images.githubusercontent.com/3467241/188790391-ddef89fc-e8dc-4e13-b5e6-8768dec28f3f.png)

**However, with try/catch, the error is gone:**
![image](https://user-images.githubusercontent.com/3467241/188790462-f17cc14a-79db-4e54-8a07-27eadb92a97c.png)

Do I miss anything?

**Reply by franklin-huang** [reply]
Somehow, if we remove `.catch` from the promise, the error is gone .

**Reply by chriskrycho** [reply]
Ah, right. This is a limitation of TypeScript's inference capabilities, unfortunately. If you produce a *single* expression, it can infer the full type correctly; otherwise, you have to explicitly specify the full type. There are two ways to do this:

1. Write `Result.ok<OkType, ErrType>(theOkValue)` and `Result.err<OkType, ErrType>(theErrValue)` when creating a `Result`.
2. Make it be a single expression, [as in this playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgJQKYGcCuAbeBfOAMyghDgHIYpNUBaEATxgAtyBuAKA5gbFTgCCsYOngBeRBzhwAhrHQAuOKKjAAdgHMA2gF0OeLqjWYyAUSgkoAWQzoZG-gimDhogHIQYAMQiY1AEwAafS5-VABjbDl+cIg1USJUGHDmIRgRGCUACkwobCUVdQ0ASjgxAD44AAUSEBFUAB4kfxkYGSU0jLg8cs4OWPj4BxhOhIks4DC1dMJgVCgCqiLipRrSeoa0LFwG0ZhAuHNLG3Q7B3LKiudCJJS9rIADAHo5dNF0J4ASBEmjGbmoHgHsVnNIAHQsIxZUHSOBZKAYSDxVClCooDA4GBgiAAa3hiLi6FQYJabWKwVhsKyqMqW0xYPmUCyR2gJzOxL2Hm8vgCIMpxU4QA), which has the two-argument `.then()` callback instead of separate `.then()` and `.catch()` callbacks.

The reason removing the `.catch()` works is that it drops into (2).

**Reply by franklin-huang** [reply]
Thanks @chriskrycho
We end up using the first approach. That would work.
